src/org/traccar/protocol/EelinkProtocolDecoder.java

```diff
@@ -131,38 +178,120 @@ private Position decodeOld(DeviceSession deviceSession, ChannelBuffer buf, int t
-        }
-
-        if (buf.readableBytes() >= 2) {
-            decodeStatus(position, buf.readUnsignedShort());
-        }
+        return position;
+    }
 
-        if (buf.readableBytes() >= 2 * 4) {
+    private Position decodeDownlink(
+            DeviceSession deviceSession, ChannelBuffer buf, Channel channel, Position position, int type, int index) {
 
-            position.set(Position.KEY_BATTERY, buf.readUnsignedShort() * 0.001);
+        getLastLocation(position, new Date());
+        position.setValid(false);
 
-            position.set(Position.KEY_RSSI, buf.readUnsignedShort());
+        switch (buf.readUnsignedByte()) {
+            case MSGSIGN_COMMAND:
+                buf.skipBytes(4); //server flag
 
-            position.set(Position.PREFIX_ADC + 1, buf.readUnsignedShort());
-            position.set(Position.PREFIX_ADC + 2, buf.readUnsignedShort());
+                int remainingbytes = buf.readableBytes();
+                String content = buf.readBytes(remainingbytes).toString(StandardCharsets.UTF_8);
 
+                position.set(Position.KEY_RESULT, content);
+                break;
+            default:
+                position = null;
+                break;
         }
 
         return position;
     }
 
-    private Position decodeNew(DeviceSession deviceSession, ChannelBuffer buf, int index) {
+    private Position decodeOld(
+            DeviceSession deviceSession, ChannelBuffer buf, Channel channel, Position position, int type, int index) {
+
+        decodeOldLocationData(deviceSession, buf, position);
+
+        switch (type) {
+            case MSG_ALARM:
+                position.set(Position.KEY_ALARM, decodeAlarm(buf.readUnsignedByte()));
+                break;
+            case MSG_STATE:
+                switch (buf.readUnsignedByte()) { // type
+                    case 0x01:
+                    case 0x02:
+                    case 0x03:
+                        buf.skipBytes(2);
+                        decodeStatus(position, buf.readShort());
+                        break;
+                    default:
+                        break;
+                }
+
+                break;
+            case MSG_GPS:
+                // only read remaining 10 bytes if they exist as they are optional in spec
+                if (buf.readableBytes() >= 2 * 5) {
+                    decodeStatus(position, buf.readUnsignedShort());
+
+                    position.set(Position.KEY_BATTERY, buf.readUnsignedShort() * 0.001);
+
+                    position.set(Position.KEY_RSSI, buf.readUnsignedShort());
+
+                    position.set(Position.PREFIX_ADC + 1, buf.readUnsignedShort());
+                    position.set(Position.PREFIX_ADC + 2, buf.readUnsignedShort());
+                }
+                break;
+            case MSG_SMS:
+                String phone = buf.readBytes(21).toString(StandardCharsets.US_ASCII);
+                String realnumber = phone.replaceAll("^(.*)\\+", "+").trim();
```

Comment:
For some reason the device puts 20 before the phone number and this is to remove it for display, its not documented in their spec but its also not part of the number at all that I am sending the message from. + will always be the start of the number and nothing should precede it.